onnxruntime/core/providers/cuda/nn/conv.cc

```diff
@@ -86,6 +93,38 @@ Status SliceOutUnwantedOutputSection(cudaStream_t stream,
   return SliceCuda::Impl(stream, input_data, input_dims, output_data, compute_metadata, element_size);
 }
 
+template <typename T, bool NHWC>
+Status Conv<T, NHWC>::PrePack(const Tensor& tensor, int input_idx, AllocatorPtr alloc,
+                              bool& is_packed, [[maybe_unused]] PrePackedWeights* prepacked_weights) {
+  is_packed = false;
+  // only layout of weight input is adjusted via PrePack
+  if (NHWC) {  // InputTensors::IN_W
+    if (input_idx == 1) {
+      // Transpose from {M, C/group, kH, kW} to {M, kH, kW, C/group}
+      auto orig_shape = tensor.Shape();
+
+      InlinedVector<size_t> perm{0, 2, 3, 1};
+      gsl::span<size_t> permutation(perm.data(), 4);
+      TensorShapeVector new_dims{orig_shape[0],
+                                 orig_shape[2],
+                                 orig_shape[3],
+                                 orig_shape[1]};
+      W_ = Tensor::Create(tensor.DataType(), TensorShape(new_dims), std::move(alloc));
+
+      auto status = cuda::Transpose::DoTranspose(GetDeviceProp(),
```

Comment:
If I understand correctly, `DefaultCudaStream()` returns [this](https://github.com/microsoft/onnxruntime/blob/f158f394d695a196ae2e06b350523a7dc741321d/onnxruntime/core/providers/cuda/cuda_execution_provider.cc#L259) which on most cases (no user provided stream and no cuda graphs) is just the default stream. This is used to do the pre-packing operation whereas the Session::Run() is using the stream provided by StreamKernelContext [here](https://github.com/microsoft/onnxruntime/blob/f158f394d695a196ae2e06b350523a7dc741321d/onnxruntime/core/framework/sequential_executor.cc#L455). Not sure how the pre-packing operation is synchronized withstream used in `Run()`. If my understanding is correct, there is a race condition here without an explicit stream sync.

Reply:
As the stream is forwarded to the constructor of `PerThreadContext` [here](https://github.com/gedoensmax/onnxruntime/blob/350f4cd18dccf288dafb5d9ca5a4a0b69eb5b9d4/onnxruntime/core/providers/cuda/cuda_execution_provider.cc#L338-L341) I was pretty sure that this is the stream that is used universal in the provider.

Reply:
That would have to be in session_state.cc then ? I will have to dive a bit deeper how to get to the stream in that function.

Reply:
How about keeping your existing changes as is but just do cudaStreamSynchronize(stream_); in PrePack() ? That should ensure solve all synchronization related issues around using different streams right ?

Reply:
Maybe i will just resort to that and see how to do this different later. It sounds like a big performance limiter.

Reply:
Just curious - why do you think it is a big perf limiter ? `PrePack()` is invoked just once per-kernel at model load time ? Synchronizing there should not affect runtime performance ?

Reply:
Due to launch latency. If we launch the transpose kernel inside pre pack we will have to wait for it to be finished. rather than already launching the next transpose. It's probably gonna be fine, but usually the paradigm would be to synchronize as little as possible, just like during an inference call.

Reply:
Whilst it will be nice to make session load a bit faster, the tradeoff to consider is to consider code change required to make that happen. Feel free to see if you can avoid the sync altogether or atleast leave behind a TODO to improve this for later.